CheMPS2/chemps2.1
.\" Manpage for chemps2 executable
.\" Contact sebastianwouters [at] gmail.com
.TH CHEMPS2 1 "9 March 2022" "version 1.8.12" "chemps2 v1.8.12"

.SH NAME
chemps2 \- spin-adapted DMRG for ab initio quantum chemistry

.SH SYNOPSIS
.B chemps2
.RI "[" "OPTION" "]"
.RI "..."

.SH DESCRIPTION
\fBchemps2\fR is a scientific code to perform spin-adapted density matrix renormalization group (DMRG) calculations for ab initio quantum chemistry fcidump files. This method allows one to obtain numerical accuracy in active spaces beyond the capabilities of full configuration interaction (FCI) and can return the active space 2-, 3-, and 4-RDM. The method is therefore ideal to replace the FCI solver in the complete active space self consistent field (CASSCF) and complete active space second order perturbation theory (CASPT2) methods. The link to the user manual can be found in the section \fBSEE ALSO\fR.

.SH OPTIONS
.SS SYMMETRY
Conventions for the symmetry group and irrep numbers (same as psi4):
.PP
.EX
                 |  0    1    2    3    4    5    6    7   
        ---------|-----------------------------------------
        0 : c1   |  A                                      
        1 : ci   |  Ag   Au                                
        2 : c2   |  A    B                                 
        3 : cs   |  Ap   App                               
        4 : d2   |  A    B1   B2   B3                      
        5 : c2v  |  A1   A2   B1   B2                      
        6 : c2h  |  Ag   Bg   Au   Bu                      
        7 : d2h  |  Ag   B1g  B2g  B3g  Au   B1u  B2u  B3u 
.EE
.SS ARGUMENTS
.TP
.BR "\-f" ", " "\-\-file=\fIinputfile\fB"
Specify the input file.
.TP
.BR "\-v" ", " "\-\-version"
Print the version of chemps2.
.TP
.BR "\-h" ", " "\-\-help"
Display this help.
.SS INPUT FILE
.TP
.BR "FCIDUMP = \fI/path/to/fcidump\fB"
Note that orbital irreps in the FCIDUMP file follow molpro convention!
.TP
.BR "GROUP = \fIint\fB"
Set the psi4 symmetry group number [0\-7] which corresponds to the FCIDUMP file.
.TP
.BR "MULTIPLICITY = \fIint\fB"
Overwrite the spin multiplicity [2S+1] of the FCIDUMP file.
.TP
.BR "NELECTRONS = \fIint\fB"
Overwrite the number of electrons of the FCIDUMP file.
.TP
.BR "IRREP = \fIint\fB"
Overwrite the target wavefunction irrep [0\-7] of the FCIDUMP file (psi4 convention).
.TP
.BR "EXCITATION = \fIint\fB"
Set which excitation should be calculated. If zero, the ground state is calculated (default 0).
.TP
.BR "SWEEP_STATES = \fIint,int,int\fB"
Set the number of reduced renormalized basis states for the successive sweep instructions (positive integers).
.TP
.BR "SWEEP_ENERGY_CONV = \fIflt,flt,flt\fB"
Set the energy convergence to stop the successive sweep instructions (positive floats).
.TP
.BR "SWEEP_MAX_SWEEPS = \fIint,int,int\fB"
Set the maximum number of sweeps for the successive sweep instructions (positive integers).
.TP
.BR "SWEEP_NOISE_PREFAC = \fIflt,flt,flt\fB"
Set the noise prefactors for the successive sweep instructions (floats).
.TP
.BR "SWEEP_DVDSON_RTOL = \fIflt,flt,flt\fB"
Set the residual norm tolerance for the Davidson algorithm for the successive sweep instructions (positive floats).
.TP
.BR "NOCC = \fIint,int,int,int\fB"
Set the number of occupied (external core) orbitals per irrep (psi4 irrep ordering).
.TP
.BR "NACT = \fIint,int,int,int\fB"
Set the number of active orbitals per irrep (psi4 irrep ordering).
.TP
.BR "NVIR = \fIint,int,int,int\fB"
Set the number of virtual (secondary) orbitals per irrep (psi4 irrep ordering).
.TP
.BR "MOLCAS_2RDM = \fI/path/to/2rdm/output\fB"
When all orbitals are active orbitals, write out the 2-RDM in HDF5 format when specified (default unspecified).
.TP
.BR "MOLCAS_3RDM = \fI/path/to/3rdm/output\fB"
When all orbitals are active orbitals, write out the 3-RDM in HDF5 format when specified (default unspecified).
.TP
.BR "MOLCAS_F4RDM = \fI/path/to/f4rdm/output\fB"
When all orbitals are active orbitals, write out the 4-RDM contracted with the Fock operator in HDF5 format when specified (default unspecified).
.TP
.BR "MOLCAS_FOCK = \fI/path/to/fock/input\fB"
When all orbitals are active orbitals, read in this file containing the Fock operator (default unspecified).
.TP
.BR "MOLCAS_FIEDLER = \fIbool\fB"
When all orbitals are active orbitals, switch on orbital reordering based on the Fiedler vector of the exchange matrix (TRUE or FALSE; default FALSE).
.TP
.BR "MOLCAS_ORDER = \fIint,int,int,int\fB"
When all orbitals are active orbitals, provide a custom orbital reordering (default unspecified). When specified, this option takes precedence over MOLCAS_FIEDLER.
.TP
.BR "MOLCAS_OCC = \fIint,int,int,int\fB"
When all orbitals are active orbitals, set initial guess to an ROHF determinant (default unspecified). The occupancy integers should be 0, 1 or 2 and the orbital ordering convention is FCIDUMP.
.TP
.BR "MOLCAS_MPS = \fIbool\fB"
When all orbitals are active orbitals, switch on the creation of MPS checkpoints (TRUE or FALSE; default FALSE).
.TP
.BR "MOLCAS_STATE_AVG = \fIbool\fB"
Switch on writing to disk of N-RDMs of intermediate roots (TRUE or FALSE; default FALSE).
.TP
.BR "SCF_STATE_AVG = \fIbool\fB"
Switch on state\-averaging (TRUE or FALSE; default FALSE).
.TP
.BR "SCF_DIIS_THR = \fIflt\fB"
Switch on DIIS when the update norm is smaller than the given threshold (default 0.0).
.TP
.BR "SCF_GRAD_THR = \fIflt\fB"
Gradient norm threshold for convergence of the DMRG\-SCF orbital rotations (default 1e\-6).
.TP
.BR "SCF_MAX_ITER = \fIint\fB"
Specify the maximum number of DMRG\-SCF iterations (default 100).
.TP
.BR "SCF_ACTIVE_SPACE = \fIchar\fB"
Rotate the active space orbitals: no additional rotations (I), natural orbitals (N), localized and ordered orbitals (L), or ordered orbitals only (F) (default I).
.TP
.BR "SCF_MOLDEN = \fI/path/to/molden\fB"
Rotate the FCIDUMP orbitals to the DMRG-SCF occupied (external core), active, and virtual (secondary) orbitals.
.TP
.BR "CASPT2_CALC = \fIbool\fB"
Switch on the CASPT2 calculation (TRUE or FALSE; default FALSE).
.TP
.BR "CASPT2_ORBS = \fIchar\fB"
Perform the DMRG calculation for the 4\-RDM in the SCF_ACTIVE_SPACE orbitals (A) or in the pseudocanonical orbitals (P) (default A).
.TP
.BR "CASPT2_IPEA = \fIflt\fB"
Ionization potential \- electron affinity shift (default 0.0).
.TP
.BR "CASPT2_IMAG = \fIflt\fB"
Imaginary level shift (default 0.0).
.TP
.BR "CASPT2_CHECKPT = \fIbool\fB"
Create checkpoints to continue the CASPT2 4\-RDM calculation over multiple runs (TRUE or FALSE; default FALSE).
.TP
.BR "CASPT2_CUMUL = \fIbool\fB"
Use a cumulant approximation for the CASPT2 4\-RDM and overwrite CASPT2_CHECKPT to FALSE (TRUE or FALSE; default FALSE).
.TP
.BR "PRINT_CORR = \fIbool\fB"
Print correlation functions (TRUE or FALSE; default FALSE).
.TP
.BR "TMP_FOLDER = \fI/path/to/tmp/folder\fB"
Overwrite the tmp folder for the renormalized operators. With MPI, separate folders per process can (but do not have to) be used (default /tmp).
.SS EXAMPLE
.PP
.EX
 $ cd /tmp
 $ wget 'https://github.com/SebWouters/CheMPS2/raw/master/tests/matrixelements/N2.CCPVDZ.FCIDUMP'
 $ ls \-al N2.CCPVDZ.FCIDUMP
 $ wget 'https://github.com/SebWouters/CheMPS2/raw/master/tests/test14.input'
 $ sed \-i "s/path\\/to/tmp/" test14.input
 $ cat test14.input
 $ chemps2 \-\-file=test14.input
.EE

.SH AUTHOR
Written by
Sebastian Wouters <\&\%sebastianwouters@gmail\&.com\&>

.SH BUGS
Reporting bugs:
\fI\%https://github.com/sebwouters/CheMPS2/issues\fR

.SH SEE ALSO
User manual:
\fI\%http://sebwouters.github.io/CheMPS2/index.html\fR

.SH COPYRIGHT
.EX
CheMPS2: a spin-adapted implementation of DMRG for ab initio quantum chemistry
Copyright (C) 2013-2021 Sebastian Wouters

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
.EE



CheMPS2/CHANGELOG.md
#### Version 1.8.12 LTS update (2022-03-09):
* Use double instead of long double in Wigner:
  max(abs(diff(wigner6j(2ja, 2jb, 2jc, 2jd, 2je, 2jf))))
    - 2j <= 95, clang : 6.90446e-12
    - 2j <= 95, gcc   : 7.01087e-12
    - 2j <= 40, clang : 2.16754e-15
    - 2j <= 40, gcc   : 1.99840e-15

#### Version 1.8.11 LTS update (2022-01-09):
* Don't require exact HDF5 version

#### Version 1.8.10 LTS update (2021-02-01):
* Fix building with HDF5 1.12

#### Version 1.8.9 LTS update (2018-10-29):
* Bug fix long double literal Wigner.cpp

#### Version 1.8.8 LTS update (2018-09-26):
* Print the number of MPS variables

#### Version 1.8.7 LTS update (2018-03-25):
* Bump up SO version from 2 to 3

#### Version 1.8.6 LTS update (2018-03-16):
* Set initial guess to an ROHF determinant (for @quanp)

#### Version 1.8.5 LTS update (2018-01-18):
* Fix build; bump up required CMake (thanks @quanp)

#### Version 1.8.4 LTS update (2017-05-23):
* MOLCAS state average CASPT2 interface (thanks @quanp)
* Update build system for psi4 (thanks @loriab)

#### Version 1.8.3 LTS update (2016-11-15):
* Custom orbital reordering for MOLCAS interface
* Fix building PyCheMPS2 (thanks @loriab)

#### Version 1.8.2 LTS update (2016-10-26):
* Fix PyCheMPS2 build

#### Version 1.8.1 LTS update (2016-10-25):
* MPS checkpoint for MOLCAS interface
* Fiedler order checkpoint for MOLCAS interface

#### Version 1.8 LTS (2016-08-24):
* Fix slow convergence linear Davidson algorithm
* Fix MPI bug binary
* MOLCAS interface binary (2-RDM, 3-RDM, and F.4-RDM dump)
* Fortran example to read in CASPT2 intermediates for MOLCAS

#### Version 1.7.3 (2016-07-14):
* DMRG workshop in sphinx documentation

#### Version 1.7.2 (2016-07-07):
* Rotate molden files with DMRG-SCF unitary from executable

#### Version 1.7.1 (2016-06-29):
* Global reordering of the active space orbitals (mixing
  irreps) based on Fiedler vector of the exchange matrix
* Davidson algorithm for linear systems (for CASPT2)

#### Version 1.7 (2016-05-24):
* Class ConjugateGradient
* Class TensorOperator
* Class ThreeDM: DMRG 3-RDM in O(L^4 D^3 + L^6 D^2)
* Class Cumulant: Cumulant reconstruction of the DMRG 4-RDM
* Class CASPT2: Internally contracted CASPT2
* Class Molden: Rotate molpro/psi4 molden to CAS orbitals
* Class Excitation: Build |1> = ( b + a * ( E_ij + E_ji )) |0>
* Part of DMRG 4-RDM via DMRG::Symm4RDM
* DMRG sweeps after RDM calculation allowed
* FCI 3-RDM in FCI::Fill3RDM
* FCI 4-RDM in FCI::Fill4RDM
* FCI diagonal 4-RDM in FCI::Diag4RDM
* FCI 4-RDM contraction with Fock operator in FCI::Fock4RDM
* Blockwise ERI rotations with DMRGSCFrotations using disk
* Update binary for DMRG-CASPT2
* Break API: bump up SO version
* Deprecate TwoDMstorage class
* Deprecate TensorDiag class
* Deprecate TensorSwap class
* Deprecate TensorS0Abase class
* Deprecate TensorS1Bbase class
* Deprecate TensorF0Cbase class
* Deprecate TensorF1Dbase class
* Deprecate TensorA class
* Deprecate TensorB class
* Deprecate TensorC class
* Deprecate TensorD class

#### Version 1.6 (2015-08-26):
* Disk i/o improvements with HDF5's hyperslab
* Performance counters in DMRG class
* Faster preconditioner FCI Green's function solver
* Bug fix FCIDUMP read-in
* chemps2 binary
* manpage for binary

#### Version 1.5 (2015-06-18):
* DMRG-CI plugin pyscf
* DMRG-SCF plugin psi4 (official release for psi4)
* Fix bug small electron number FCI class
* FCIDUMP file support
* Sphinx documentation
* DMRG class supports 4-fold permutation symmetry (i.o. 8-fold)
* Hybrid MPI & OpenMP for DMRG + 2-RDM (not for DMRG-SCF yet)

#### Version 1.4 (2014-11-23):
* 2-RDM storage class
* Optimization OpenMP over symmetry blocks Heff
* DIIS acceleration DMRG-SCF
* Two-orbital mutual information and correlation functions
* Augmented Hessian NR Edmiston-Ruedenberg orbital localization
* PyCheMPS2: python interface to libchemps2
* FCI Green's function solver
* State-averaged DMRG-SCF

#### Version 1.0 (2014-04-08):
* Initial release



QC-DMET/src/chemps2.py
'''
    QC-DMET: a python implementation of density matrix embedding theory for ab initio quantum chemistry
    Copyright (C) 2015 Sebastian Wouters
    
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.
    
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
    
    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
'''

import numpy as np
import ctypes
import os  # for dev/null
import sys # for sys.stdout
import qcdmet_paths
import PyCheMPS2

def solve( CONST, OEI, FOCK, TEI, Norb, Nel, Nimp, chempot_imp=0.0, printoutput=False ):

    # Set the seed of the random number generator and cout.precision
    Initializer = PyCheMPS2.PyInitialize()
    Initializer.Init()

    # Setting up the Hamiltonian
    Group = 0
    orbirreps = np.zeros([ Norb ], dtype=ctypes.c_int)
    HamCheMPS2 = PyCheMPS2.PyHamiltonian(Norb, Group, orbirreps)
    HamCheMPS2.setEconst( CONST )
    for cnt1 in range(Norb):
        for cnt2 in range(Norb):
            HamCheMPS2.setTmat(cnt1, cnt2, FOCK[cnt1, cnt2])
            for cnt3 in range(Norb):
                for cnt4 in range(Norb):
                    HamCheMPS2.setVmat(cnt1, cnt2, cnt3, cnt4, TEI[cnt1, cnt3, cnt2, cnt4]) #From chemist to physics notation
    '''HamCheMPS2.save()
    exit(123)'''
    if (chempot_imp != 0.0):
        for orb in range(Nimp):
            HamCheMPS2.setTmat(orb, orb, FOCK[orb, orb] - chempot_imp)
    
    # Killing output if necessary
    if ( printoutput==False ):
        sys.stdout.flush()
        old_stdout = sys.stdout.fileno()
        new_stdout = os.dup(old_stdout)
        devnull = os.open('/dev/null', os.O_WRONLY)
        os.dup2(devnull, old_stdout)
        os.close(devnull)
        
    if ( Norb <= 10 ):
    
        # FCI ground state calculation
        assert( Nel % 2 == 0 )
        Nel_up       = Nel / 2
        Nel_down     = Nel / 2
        Irrep        = 0
        maxMemWorkMB = 100.0
        FCIverbose   = 2
        theFCI = PyCheMPS2.PyFCI( HamCheMPS2, Nel_up, Nel_down, Irrep, maxMemWorkMB, FCIverbose )
        GSvector = np.zeros( [ theFCI.getVecLength() ], dtype=ctypes.c_double )
        theFCI.FillRandom( theFCI.getVecLength() , GSvector ) # Random numbers in [-1,1[
        GSvector[ theFCI.LowestEnergyDeterminant() ] = 12.345 # Large component for quantum chemistry
        EnergyCheMPS2 = theFCI.GSDavidson( GSvector )
        #SpinSquared = theFCI.CalcSpinSquared( GSvector )
        TwoRDM = np.zeros( [ Norb**4 ], dtype=ctypes.c_double )
        theFCI.Fill2RDM( GSvector, TwoRDM )
        TwoRDM = TwoRDM.reshape( [Norb, Norb, Norb, Norb], order='F' )
        TwoRDM = np.swapaxes( TwoRDM, 1, 2 ) #From physics to chemistry notation
        del theFCI
        
    else:
    
        # DMRG ground state calculation
        assert( Nel % 2 == 0 )
        TwoS  = 0
        Irrep = 0
        Prob  = PyCheMPS2.PyProblem( HamCheMPS2, TwoS, Nel, Irrep )

        OptScheme = PyCheMPS2.PyConvergenceScheme(3) # 3 instructions
        #OptScheme.setInstruction(instruction, D, Econst, maxSweeps, noisePrefactor)
        OptScheme.setInstruction(0,  500, 1e-10,  3, 0.05)
        OptScheme.setInstruction(1, 1000, 1e-10,  3, 0.05)
        OptScheme.setInstruction(2, 1000, 1e-10, 10, 0.00) # Last instruction a few iterations without noise

        theDMRG = PyCheMPS2.PyDMRG( Prob, OptScheme )
        EnergyCheMPS2 = theDMRG.Solve()
        theDMRG.calc2DMandCorrelations()
        TwoRDM = np.zeros( [Norb, Norb, Norb, Norb], dtype=ctypes.c_double )
        for orb1 in range(Norb):
            for orb2 in range(Norb):
                for orb3 in range(Norb):
                    for orb4 in range(Norb):
                        TwoRDM[ orb1, orb3, orb2, orb4 ] = theDMRG.get2DMA( orb1, orb2, orb3, orb4 ) #From physics to chemistry notation

        # theDMRG.deleteStoredMPS()
        theDMRG.deleteStoredOperators()
        del theDMRG
        del OptScheme
        del Prob

    del HamCheMPS2

    # Reviving output if necessary
    if ( printoutput==False ):
        sys.stdout.flush()
        os.dup2(new_stdout, old_stdout)
        os.close(new_stdout)
    
    # Calculate the energy of the impurity based on the 1RDM and 2RDM
    OneRDM = np.einsum( 'ijkk->ij', TwoRDM ) / ( Nel - 1 )
    
    # To calculate the impurity energy, rescale the JK matrix with a factor 0.5 to avoid double counting: 0.5 * ( OEI + FOCK ) = OEI + 0.5 * JK
    ImpurityEnergy = CONST
    ImpurityEnergy += 0.5 * np.einsum( 'ij,ij->', OneRDM[:Nimp,:], OEI[:Nimp,:] + FOCK[:Nimp,:] )
    ImpurityEnergy += 0.5 * np.einsum( 'ijkl,ijkl->', TwoRDM[:Nimp,:,:,:], TEI[:Nimp,:,:,:] )
    return ( ImpurityEnergy, OneRDM )

